keps/sig-cli/20190920-future-of-kubectl-cp.md

```diff
@@ -0,0 +1,141 @@
+provides transparency when copying files as well as more proven mitigations for path traversals, symlink directory escapes, tar bombs, and other exploits.
+Use of tar is also more more featureful, in that it can preserve file permissions and copy pod-to-pod.   
+
+With all of this in mind the cost and risk of maintaining the cp command should be weighed against what is considered crucial functionality in kubectl. 
+It's better to address 80% of use cases with a simple tool than trying to address the remaining 20% at the cost of risking those 80%.     
+
+[1] https://github.com/kubernetes/kubernetes/pull/78622   
+[2] https://github.com/kubernetes/kubernetes/pull/73053   
+[3] https://gist.github.com/tallclair/9217e2694b5fdf27b55d6bd1fda01b53   
+
+## Motivation
+
+- The `kubectl cp` command has been the subject of multiple security vulnerability reports.
+- Even with basic use cases of `kubectl cp`, there are long-standing issues that have not been addressed.   
+     1) container images are required to have the tar binary in order to use `kubectl cp`.    
+     2) cp command is not well maintained for Windows, if it works at all.    
+
+Given its history, `kubectl cp` should perform only basic copying of files.  Advanced features of file copying should be out of scope for `kubectl cp`.  
+
+### Goals for kubectl cp
+
+- Provide simple function to copy a single file or directory, without advanced features such as symlinks or file permission changes
+- Don't accept new features or flags to cp beyond the use cases listed in the user stories below.
+- Offer users example `kubectl exec`/ shell commands to address advanced options of copying files. (There are a few already in --help)
+
+### Non-Goals
+
+- Rewrite `kubectl cp` to not use tar, by modifying CRI as outlined partially [here](https://github.com/kubernetes/kubernetes/issues/58512). 
+
+## Proposal
+
+- `kubectl cp` should provide simple function of copying single file or directory between local environments and pods.
+- Identify and document `kubectl exec` commands to address more advanced options for copying files.  
+- Only the user stories listed below should be supported by `kubectl cp`.  
+
+### User Stories
+
+#### UserA wants to copy a file or a directory from a local host to a pod (or a container in a pod)
+
+```console
+ $ kubectl cp localdir somens/somepod:/poddir (-c acontainer)
+ $ kubectl cp localdir somens/somepod:/poddir (-c acontainer)
+ $ kubectl cp localdir/filea somens/somepod:/poddir (-c acontainer)
+```
+
+#### UserB wants to copy a file or a directory from a pod to a local host
+
+```console
+ $ kubectl cp somens/somepod:/poddir localdir
+ $ kubectl cp somens/somepod:/poddir/filea localdir/filea
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+## Design Details
+
+### Test Plan
+
+### Graduation Criteria
+
+### Upgrade / Downgrade Strategy
```

Comment:
This needs to be expanded to describe how you will deprecate the existing functionality (if indeed that is the right decision)

Reply:
The process will simple, we have to remove that functionality and inform users of alternatives, which is what we've been doing on every CVE fix. My proposal is: user is getting information that file is a symlink and we will skip this file, along with detailed information how to use `kubectl exec` as a replacement. 
@sallyom mind adding something like that here

Reply:
As above, agree it's easy, but it needs to be filled in.

Reply:
i've added info there